sympy/matrices/matrices.py

```diff
@@ -3912,35 +3974,65 @@ def LUdecomposition_Simple(self, iszerofunc=_iszero):
         LUdecompositionFF
         LUsolve
         """
-        if not self.is_square:
-            raise NonSquareMatrixError(
-                "A Matrix must be square to apply LUdecomposition_Simple().")
-        n = self.rows
-        A = self.as_mutable()
-        p = []
-        # factorization
-        for j in range(n):
-            for i in range(j):
-                for k in range(i):
-                    A[i, j] = A[i, j] - A[i, k] * A[k, j]
-            pivot = -1
-            for i in range(j, n):
-                for k in range(j):
-                    A[i, j] = A[i, j] - A[i, k] * A[k, j]
-                # find the first non-zero pivot, includes any expression
-                if pivot == -1 and not iszerofunc(A[i, j]):
-                    pivot = i
-            if pivot < 0:
-                # this result is based on iszerofunc's analysis of the possible pivots, so even though
-                # the element may not be strictly zero, the supplied iszerofunc's evaluation gave True
-                raise ValueError("No nonzero pivot found; inversion failed.")
-            if pivot != j:  # row must be swapped
-                A.row_swap(pivot, j)
-                p.append([pivot, j])
-            scale = 1 / A[j, j]
-            for i in range(j + 1, n):
-                A[i, j] = A[i, j] * scale
-        return A, p
+
+        if self.rows == 0 or self.cols == 0:
+            # Define LU decomposition of a matrix with no entries as a matrix of the same dimensions with all zero
+            # entries.
+            return self.zeros(self.rows, self.cols), []
+
+        def next_pivot_f(matrix, row, col):
+
+            return nextpivotfunc(matrix, row, col, iszerofunc)
+
+        lu = self.as_mutable()
+        row_swaps = []
+
+        pivot_col = 0
+        for pivot_row in range(0, lu.rows-1):
+
+            pivot_row_cand, pivot_col = next_pivot_f(lu, pivot_row, pivot_col)
+            if pivot_col == lu.cols:
+                return lu, row_swaps
+
+            if pivot_row != pivot_row_cand:
+                # Row swap book keeping:
+                # Record which rows were swapped.
+                # Update stored portion of L factor by multiplying L on the left and right with the current
+                # permutation.
+                # Swap rows of U.
+                row_swaps.append([pivot_row, pivot_row_cand])
+
+                # Update L.
+                for col in range(0, pivot_row):
+                    tmp = lu[pivot_row, col]
+                    lu[pivot_row, col] = lu[pivot_row_cand, col]
+                    lu[pivot_row_cand, col] = tmp
```

Comment:
It seems that these three lines could be replaced by a single one using python's swap idiom: `a, b = b, a`.